### PR TITLE
Addon Docs: Remove `react` peer dependency

### PR DESCRIPTION
**Pull request by JReinhold**
Works on https://github.com/storybookjs/storybook/issues/25179

This PR is an investigation into the possibility of removing React as a peer dependency of `addon-docs`, originally proposed at https://github.com/storybookjs/storybook/discussions/24823.

### Objective

Remove `react` and `react-dom` as a peer dependency of `addon-docs` - or alternatively turn into an optional peer dependency.

We want to enable non-React users to use `addon-docs` without needing to install React themselves and “bloating” their packages. React will no doubt remain in `node_modules` as a transitive dependency, but that still feels nicer.

### Questions we want to answer

- Can we supply a default instance of React with `addon-docs`, while still allowing React users to use their own instance of React?

👆is important, because React users need to be able to define which version of React they want to use for rendering their components and stories.

## Findings

We’ve found a strategy to supply `react` and `react-dom` by default with `addon-docs` (via a direct dependency) while still allowing users to supply their own instance and version by simply installing the packages in the root project if they want to. We’ve proved that this strategy works in almost[^pnp-non-react] all situations.

This way, if the user does not have `react` installed, `react@18.2.0` from `addon-docs` will be used.

The strategy is to attempt to get the path to `react` using `require.resolve('react')`.

1. We try to resolve it in `@storybook/core-server` and `@storybook/react` presets, which will either get the user’s root `react` dependency, or fail. It will never resolve to the `react` in `@storybook/addon-docs`
2. If we get any paths from above, we store them as a `resolvedReact` property on presets.
3. Then in `addon-docs` we get that `resolvedReact` preset and if it’s empty (because the user didn’t have `react` in the root) we do the same resolution which will always resolve to the `react` that `addon-docs` has.
4. Then in `addon-docs` `webpackFinal` and `viteFinal` we alias all imports of `react`, `react-dom` and `@mdx-js/react`[^mdx-js] to the paths defined in `resolvedReact`.
5. Now all imports of `react` are either from the root or from `addon-docs`.

## Remaining Work

This PR is just a spike that proves this can work, there's more work to be done for this to be releasable.

See https://github.com/storybookjs/storybook/issues/25179

- ~Find out why the unit test is failing, maybe the React import needs aliased as well~
- ~Update documentation to not say React is necessary - if we do that~
- ~Update documentation to note that users can specify their own React version for docs if they need to~
- ~Automigration to remove React as a dependency if possible - there might be a lot of edge cases here making this "if possible" detection very hard~
- ~Stop adding React dependencies on `init`~
- ~Test in all permutations of builders package managers and operating systems~
- ~We might want to make it possible to _opt-out_ of this, because there could be unknown unknowns where this strategy brakes users' setups~

### E2E Test

I propose that we add an E2E test that verifies that React is available in MDX.

1. Create an MDX file that imports `version` from `react` and displays it
2. Create a React component that does the same, and include it in the MD file
3. Add an E2E test to addon docs that visits this MDX doc and asserts the React version is rendered as expected.

### Manual Test

Before merging this we should manually test it in multiple project permutations:

| **Renderer** | **Builder** | **Package manager** | **OS** | **Result** |
|---|---|---|---|---|
| React | Vite | npm | macOS | ✅ |
| React | Vite | Yarn `node_modules` | macOS | ✅ |
| React | Vite | Yarn PnP | macOS | ✅ |
| React | Vite | pnpm | macOS | ✅ |
| React | Webpack | npm | macOS | ✅ |
| React | Webpack | Yarn `node_modules` | macOS | ✅ |
| React | Webpack | Yarn PnP | macOS | ✅ |
| React | Webpack | pnpm | macOS | ✅ |
| Vue3 | Vite | npm | macOS | ✅ |
| Vue3 | Vite | Yarn `node_modules` | macOS | ✅ |
| Vue3 | Vite | Yarn PnP | macOS | ✅ |
| Vue3 | Vite | pnpm | macOS | ✅ |
| Vue3 | Webpack | npm | macOS | ✅ |
| Vue3 | Webpack | Yarn `node_modules` | macOS | ✅ |
| Vue3 | Webpack | Yarn PnP | macOS | ✅ |
| Vue3 | Webpack | pnpm | macOS | ✅ |
| Vue3+React 17 | Vite | npm | macOS | ✅ |
| Vue3+React 17 | Vite | Yarn `node_modules` | macOS | ✅ |
| Vue3+React 17 | Vite | Yarn PnP | macOS | ❌ |
| Vue3+React 17 | Vite | pnpm | macOS | ✅ |
| Vue3+React 17 | Webpack | npm | macOS | ✅ |
| Vue3+React 17 | Webpack | Yarn `node_modules` | macOS | ✅ |
| Vue3+React 17 | Webpack | Yarn PnP | macOS | ❌ |
| Vue3+React 17 | Webpack | pnpm | macOS | ✅ |
| React | Vite | npm | Windows | ✅ |
| React | Vite | Yarn PnP | Windows | ✅ |
| SvelteKit | Vite | npm | Windows | ✅ |
| SvelteKit | Vite | Yarn PnP | Windows | ❌ |
| React | Webpack | npm | Windows | ✅ |
| React | Webpack | Yarn PnP | Windows | ✅ |
| Vue3 | Webpack | npm | Windows | ✅ |
| Vue3 | Webpack | Yarn PnP | Windows | ✅ |
| Vue3+React 17 | Webpack | npm | Windows | ✅ |
| Vue3+React 17 | Webpack | Yarn PnP | Windows | ✅ |

[^pnp-non-react]: This doesn’t work in the specific combination of Yarn PnP + Non-React framework + user having its own React version in the root package. It doesn’t break, but it will always use `react@18.2.0` from `addon-docs` and ignore the `react` in the root. We think this is okay because:

	- It’s only a problem in Yarn PnP
	- It’s only a problem in Non-React setups
	- It’s only a problem if the user wants to supply their own version of React

	Bullet 2 and 3 seems highly unlikely to appear at the same time, why would a Svelte user care about which React version their docs are using?

	It has a workaround, the user can set the `resolvedReact` property in `main.ts` manually with `require.resolve('react')`

This pull request has been released as version [`0.0.0-pr-24881-sha-6ea23655`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24881-sha-6ea23655). Install it by pinning all your Storybook dependencies to that version.
  <details>
  <summary>More information</summary>
  
  | | |
  | --- | --- |
  | **Published version** | [`0.0.0-pr-24[8](https://github.com/storybookjs/storybook/actions/runs/7207572473/job/19634682139#step:10:8)81-sha-6ea23655`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24881-sha-6ea23655) |
  | **Triggered by** | @JReinhold |
  | **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
  | **Branch** | [`addon-docs-without-react`](https://github.com/storybookjs/storybook/tree/addon-docs-without-react) |
  | **Commit** | [`6ea23655`](https://github.com/storybookjs/storybook/commit/6ea2365560ebe808f2b3[9](https://github.com/storybookjs/storybook/actions/runs/7207572473/job/19634682139#step:10:9)6ab152ef0666b7cc780) |
  | **Datetime** | Thu Dec 14 [10](https://github.com/storybookjs/storybook/actions/runs/7207572473/job/19634682139#step:10:10):26:44 UTC 2023 (`1702549604`) |
  | **Workflow run** | [7207572473](https://github.com/storybookjs/storybook/actions/runs/7[20](https://github.com/storybookjs/storybook/actions/runs/7207572473/job/19634682139#step:10:20)7572473) |
  
  To request a new release of this pull request, mention the `@storybookjs/core` team.
  
  _core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=[24](https://github.com/storybookjs/storybook/actions/runs/7207572473/job/19634682139#step:10:24)881`_
  </details>

[^mdx-js]: Webpack has a weird behavior with `@mdx-js/react` and symlinked stories like our template-stories in the monorepo. If the user has a custom React version installed, `@mdx-js/react` will not be hoisted from `addon-docs`, and then Webpack can't find it when rendering symlinked MDX files. So we explicitly alias it to fix this - this doesn't seem to be necessary in Vite projects.
[^nextjs]: Our Next.js integration already aliases `react` and `react-dom` to the one in `next/compiled`, so none of this work has an affect on Next.js, and the user cannot define their own version of React. This just means that the problem here is already fixed for Next.js users.

### CODE REVIEW

File: code/lib/core-server/src/presets/common-preset.ts [hunk 1]
```diff
@@ -349,3 +349,22 @@ export const experimental_serverChannel = async (
...
+ * If the user doesn't have an explicit dependency on react this will return the existing values
+ * Which will be the versions shipped with addon-docs
+ */
+export const resolvedReact = async (existing: any) => {
+  try {
+    return {
+      ...existing,
+      react: dirname(require.resolve('react/package.json')),
+      reactDom: dirname(require.resolve('react-dom/package.json')),
+      mdx: dirname(require.resolve('@mdx-js/react/package.json')),
```
Comment (valentinpalkovic) [inline code/lib/core-server/src/presets/common-preset.ts:365]: Since `react`, `react-dom` and `@mdx-js` are not defined as any kind of dependencies in `@storybook/core-server`, this will fail in yarn pnp environments.

File: code/renderers/react/src/preset.ts [hunk 2]
```diff
@@ -16,3 +16,26 @@ export const previewAnnotations: StorybookConfig['previewAnnotations'] = async (
     .concat([join(__dirname, 'entry-preview.mjs')])
     .concat(docsEnabled ? [join(__dirname, 'entry-preview-docs.mjs')] : []);
 };
+
+/**
+ * Try to resolve react and react-dom from the root node_modules of the project
+ * addon-docs uses this to alias react and react-dom to the project's version when possible
+ * If the user doesn't have an explicit dependency on react this will return the existing values
```
Comment (valentinpalkovic) [inline code/renderers/react/src/preset.ts:23]: I don't get this. The user must directly depend on `react` if `@storybook/react` is in use or do I oversee something?

File: code/lib/core-server/src/presets/common-preset.ts [hunk 3]
```diff
@@ -349,3 +349,22 @@ export const experimental_serverChannel = async (
...
+    };
+  } catch (e) {
+    return existing;
+  }
+};
```
Comment (valentinpalkovic) [inline code/lib/core-server/src/presets/common-preset.ts:370]: I don't understand why we need this (Also, it shouldn't work in pnp environments). So either we get the react, react-dom and mdx package resolved from `@storybook/react` or from `@storybook/addon-docs`. Why must we do the same lookup in `@storybook/core-server`?

**Comment by valentinpalkovic** [comment]
Well done Jeppe!

**Comment by ndelangen** [comment]
@SocketSecurity ignore @vitejs/plugin-vue@4.4.0
@SocketSecurity ignore json-parse-even-better-errors@3.0.0
@SocketSecurity ignore ignore-walk@6.0.3
@SocketSecurity ignore json-stable-stringify@1.0.2
@SocketSecurity ignore @babel/preset-modules@0.1.6
@SocketSecurity ignore piscina@4.0.0
@SocketSecurity ignore flat-cache@3.1.0

**Comment by ndelangen** [comment]
Looks like this work is conflicting with the recent work to support RSC, @shilman

**Review by ndelangen — APPROVED** [review]

**Comment by JReinhold** [comment]
Tested this manually across all the project permutations, and everything works as expected. The cases that don't work are already known.

🎉

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
@JReinhold merge?
